[llvm] Update to version 10.0.0 and add new features

yurybura opened the pull request:
LLVM 10 release is coming soon.
So, this PR updates LLVM to 10.0.0 version.
Other changes:
- use Github monorepo;
- support sub-projects compiler-rt, clang-tools-extra, lld, openmp, polly;
- exclude tests and examples from build to reduce disk space usage;
- enable RTTI by default

Related to #8521

Fixes #3377
Fixes #4277
Fixes #4461
Fixes #5153
Fixes #8347
Fixes #8463
Fixes #9670

yurybura committed:
[llvm] Update to version 10.0.0 and new features

yurybura committed:
switch to llvmorg-10.0.0-rc3, allow to use bin directory for tools

yurybura committed:
Fix CMake targets after vcpkg_fixup_cmake_targets call.

yurybura committed:
[halide] Update to master version to support LLVM 10

yurybura commented:
Dear @PhoebeHui, could you help to solve Windows build-bot issue?
Linker generates out of memory errors:

LINK : fatal error LNK1201: error writing to program database 'C:\vsts\_work\4\s\buildtrees\llvm\x64-windows-dbg\bin\llvm-dwp.pdb'; check for insufficient disk space, invalid path, or insufficient privilege

LINK : fatal error LNK1102: out of memory

LLVM is really big project. I can disable some features by default, but it isn't right way I think.

Thank you!

yurybura committed:
[halide] fix x64-osx build

yurybura committed:
[llvm] add enable-assertions feature

yurybura committed:
[llvm] fix comment

PhoebeHui commented:
@strega-nil, could you help take a look at above issue?

PhoebeHui commented:
@yurybura, thanks for your contribution!

Noted that the following triplets disabled on https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt#L1002, if you fix the issue with specific triplet, please remove it.

llvm:arm64-windows=fail
llvm:arm-uwp=fail
llvm:x64-uwp=fail
llvm:x64-linux=ignore
# installing iconv makes building llvm fail; needs to be fixed
llvm:x64-osx=ignore

yurybura committed:
[llvm] enable ci for x64-linux and x64-osx

yurybura commented:
@PhoebeHui, let me try to fix failed CI checks.
Thank you!

yurybura commented:
@PhoebeHui, I've already tested my changes for x64-windows and x64-linux triplets on my local machine.

Unfortunately I cannot test with CI bot...
After enable x64-linux CI, the build fails with the same problem (low amount of memory for link):
```
collect2: fatal error: ld terminated with signal 9 [Killed]
compilation terminated. 
```

PhoebeHui commented:
llvm is a complicated and big project, it brings a stress to CI test machines, @strega-nil, any suggestions about the issue?

JackBoosY commented:
This PR is awesome, thanks for this PR!

strega-nil commented:
@PhoebeHui huh! Well, that's unfortunate. Lemme think about this!

BillyONeal commented on ports/llvm/portfile.cmake:
```diff
@@ -1,132 +1,130 @@
     OUT_SOURCE_PATH SOURCE_PATH
-    ARCHIVE ${ARCHIVE}
+    REPO llvm/llvm-project
+    REF llvmorg-10.0.0-rc3
```
Is this really updating to an RC version rather than a release version?

yurybura replied:
@BillyONeal
Yes, but I hope it will be fixed soon (now RC4 is available). I wanted to start review as soon as possible because it may take long time (~weeks). Can you have a look on other changes?

BillyONeal replied:
My cmake-fu isn't good enough for me to feel confident reviewing this; I'm new on this project and wouldn't want to say much about it. @strega-nil just asked me to review this and I pointed out the only thing that looked strange to me. :)

yurybura replied:
Resolved.

d-winsor commented on ports/llvm/portfile.cmake:
```diff
@@ -1,132 +1,130 @@
+# So, rename subfolder `/tools/${PORT}` to `/bin` back because there is no way to avoid this in vcpkg_fixup_cmake_targets.
+foreach(_target IN LISTS _release_targets _debug_targets)
+    file(READ ${_target} _contents)
+    string(REPLACE "{_IMPORT_PREFIX}/tools/${PORT}" "{_IMPORT_PREFIX}/bin" _contents "${_contents}")
```
I am not familiar with vcpkg but I enlisted in the PR to try out llvm.10 and ran into an issue in Debug. The right files exist in **vcpkg\installed\x64-windows\bin** for release but there are none in **vcpkg\installed\x64-windows\debug\bin** for debug. My guess is that is happening here?

d-winsor replied:
I imagine this line should be different for debug as can be seen in the old changes on line 98. Maybe something like:

```cmake
set(_bin_prefix)
if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
    set(_bin_prefix bin)
...
if(NOT DEFINED VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
    set(_bin_prefix debug/bin)
...
string(REPLACE "{_IMPORT_PREFIX}/tools/${PORT}" "{_IMPORT_PREFIX}/${_bin_prefix}" _contents "${_contents}")
```

yurybura replied:
@d-winsor, Thank you for comments!
This version of port (like the old one) doesn't install binaries for debug configuration. Debug binaries are removed at line 118. Do you really need debug version of tools? Let me know how are you going to use them.
Thank you.

d-winsor replied:
The bin needs to be installed for the debug version however it is no longer being copied to `${CURRENT_PACKAGES_DIR}/debug/tools/llvm` and have the cmake files updated to reference this path. Instead they remain in the bin are deleted in the REMOVE_RECURSE call below.

The absence of these binaries causes the find_package to fail because it can't find the *.dlls. I was able to resolve this by removing the `${CURRENT_PACKAGES_DIR}/debug/bin` from the REMOVE_RECURSE and adding the line below:

```cmake
string(REPLACE "{_IMPORT_PREFIX}/tools/${PORT}" "{_IMPORT_PREFIX}/bin" _contents "${_contents}")
string(REPLACE "{_IMPORT_PREFIX}/tools/debug/${PORT}" "{_IMPORT_PREFIX}/debug/bin" _contents "${_contents}")
```

yurybura replied:
@d-winsor, Could you provide your test example? I've tested this port with Hilide only. I think debug version of llvm-config will report incorrect paths for include at least.
Thank  you.

d-winsor replied:
That paths for includes/libraries/bin are all correct it just that the content in /debug/bin is being deleted and the paths in `share/clang/**-debug.cmake` files are of the form `{_IMPORT_PREFIX}/tools/debug/${PORT}` not `{_IMPORT_PREFIX}/tools/${PORT}`

I can't share the sample I am testing on but something as simple as should hit the missing file error:
```cpp
cmake_minimum_required(VERSION 3.16)
find_package(LLVM CONFIG REQUIRED)
find_package(Clang CONFIG REQUIRED)
project(test)
add_executable(test test.cpp)
target_link_libraries(test PRIVATE clangAST)
```

yurybura replied:
@d-winsor, Thank you for report this. Let me check.

yurybura replied:
@d-winsor, It should works now. Please check.

d-winsor replied:
I believe there is still a problem here. The format of the paths in the `*-debug.cmake` files is different from `*-release.cmake` files. You should be replacing the following for debug instead:
```cmake
{_IMPORT_PREFIX}/debug/tools/${PORT}
```

yurybura replied:
Sorry for this (I can test it only tomorrow). Let me fix it shortly.

d-winsor replied:
No problems. I really appreciate you making this contribution, it is really useful to my current work.

yurybura replied:
I've added `d` postfix to the debug shared libraries and I've copied them to the ./bin directory, Now llvm-config.exe and LLVM CMake scripts should works.
@d-winsor Please let me know if you will have any trouble.

JackBoosY commented:
Note the other paths in the cmake file.
Related: #9622.

d-winsor commented on ports/llvm/portfile.cmake:
```diff
@@ -1,132 +1,130 @@
     PREFER_NINJA
-    OPTIONS ${FEATURE_OPTIONS}
-        -DLLVM_TARGETS_TO_BUILD=X86
-        -DLLVM_ABI_BREAKING_CHECKS=FORCE_OFF
```
Why was this removed? Will this not cause a linker errors in debug (as this defaults to WITH_ASSERTS)?

Atleast this is what I see when I try an link c++14 (llvm) libraries with c++17 which should be ABI compatiable.

```
... error LNK2038: mismatch detected for 'LLVM_ENABLE_ABI_BREAKING_CHECKS': value '1' doesn't match value '0' in ...
```

yurybura replied:
d-winsor, I followed the LLVM's release script: https://github.com/llvm/llvm-project/blob/release/10.x/llvm/utils/release/build_llvm_package.bat

I'll add new feature for control LLVM_ABI_BREAKING_CHECKS...

yurybura replied:
Resolved.

d-winsor commented on ports/llvm/portfile.cmake:
```diff
@@ -1,132 +1,130 @@
-        -DLLVM_TARGETS_TO_BUILD=X86
-        -DLLVM_ABI_BREAKING_CHECKS=FORCE_OFF
-        -DLLVM_TOOLS_INSTALL_DIR=tools/llvm
-        -DLLVM_PARALLEL_LINK_JOBS=1
```
Even my 16 core 3950x with 32GB of ram restarts the build becuase of memory overuse. Consider leaving this which may also resolve issues with the CI.
```
-- Restarting Build without parallelism because memory exceeded
```

yurybura replied:
@d-winsor

> ```
> -- Restarting Build without parallelism because memory exceeded
> ```
As I understand correctly, to avoid this error I should call `vcpkg_install_cmake(DISABLE_PARALLEL)`.

But OK, let me try to disable parallel linking in the LLVM.

yurybura replied:
Resolved. But LINK : fatal error LNK1201 still present on CI.

yurybura committed:
[llvm] update to v10.0.0-rc5
[llvm] disable assertions and ABI breaking checks by default
[llvm] limit the maximum number of concurrent link jobs to 1
[llvm] add postfix `d` and debug shared libraries in the `/bin` directory

yurybura committed:
[llvm] revert debug postfix `d`
[llvm] don't remove debug version of tools and shared libs

yurybura committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into llvm10

yurybura committed:
[llvm] fix debug tools path

Voskrese commented:
2 days ago · LLVM 10.0.0 Release

vcpkg_check_linkage(ONLY_STATIC_LIBRARY) 

> why only static build? after, other packages will have errors when linking

dynamic & static

https://github.com/vovkos/llvm-package-windows/releases
https://github.com/ziglang/zig/wiki/How-to-build-LLVM,-libclang,-and-liblld-from-source
https://github.com/llvm/llvm-project/releases

JackBoosY commented:
Any news?

yurybura committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into llvm10

yurybura committed:
[llvm] update to release tag llvmorg-10.0.0
[llvm] build with CMAKE_DEBUG_POSTFIX=d
[llvm] remove debug binaries

yurybura commented:
> 2 days ago · LLVM 10.0.0 Release
> 
> vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
> 
> > why only static build? after, other packages will have errors when linking
> 
> dynamic & static
> 
> https://github.com/vovkos/llvm-package-windows/releases
> https://github.com/ziglang/zig/wiki/How-to-build-LLVM,-libclang,-and-liblld-from-source
> https://github.com/llvm/llvm-project/releases

@Voskrese, thank you for your comment.

Shared libraries don't support on Windows:
https://github.com/llvm/llvm-project/blob/master/llvm/CMakeLists.txt#L567

Let's finish this PR for only_static configuration. Then let me try to enable shared libraries (at least for non-windows triplets) in a separate PR.

yurybura committed:
remove new CMake argument FOLLOW_SYMLINK_CHAIN

yurybura commented:
@JackBoosY I've almost done. Windows CI still doesn't work. Can somebody have have a look?

JackBoosY commented:
The reason we have these regressions is because the disks of the pipeline machines are full.
So don't worry about that, we will rerun Windows PR test after resolve this problem.

dan-shaw commented:
@yurybura @JackBoosY Yes, I will be removing `buildtrees/debug-folder` and `buildtrees/rel-folder` after the debug and release installs respectively so there will be more space. Will get this change in during the weekend

yurybura commented:
@JackBoosY From my side this PR is ready to merge.
Thank you.

d-winsor commented on ports/llvm/portfile.cmake:
```diff
@@ -1,139 +1,166 @@
+    file(INSTALL ${_debug_shared_libs} DESTINATION ${CURRENT_PACKAGES_DIR}/bin)
+
+    file(REMOVE_RECURSE
+        ${CURRENT_PACKAGES_DIR}/debug/bin
```
As per our previous discussion you are still deleting the bin folder after correcting to the files above to point to it. Is this a mistake or is it intentional and I am missing something about Vcpkg.

yurybura committed:
[llvm] force TableGen to be built with optimization

yurybura committed:
[llvm] add future disable-clang-static-analyzer

yurybura commented:
Hello @dan-shaw, @JackBoosY. Please have a look this PR.
Can I temporary add failed configs to the ci.baseline.txt?

JackBoosY commented:
@yurybura Since llvm is dependent on other ports, I think we can't skip llvm's pipeline test for the time being.

JackBoosY commented:
llvm test successful on x86-windows.

PhoebeHui commented:
The failures with x64-windows and x64-windows-static:

oleaut32.lib uuid.lib comdlg32.lib advapi32.lib 
LINK : fatal error LNK1201: error writing to program database 'C:\vsts\_work\4\s\buildtrees\llvm\x64-windows-dbg\bin\llvm-isel-fuzzer.pdb'; check for insufficient disk space, invalid path, or insufficient privilege
ninja: build stopped: subcommand failed.

@dan-shaw, could help have a look again? it seems need more space on CI test machine.

piotrek-szczygiel commented:
When will I be able to download this version from vcpkg repository?

ZeeWanderer commented:
COnsider adding all possible LLVM_TARGETS_TO_BUILD as a list of options. Port produces ~80 GB of artifacts on build of llvm[core,utils,tools]:x64-windows while if only X86 backend is built (both Debug and Release) ~28 GB of artifacts are produced. Did not measure compile time but it should be a lot less, thou current list of port options is already quite extensive.

yurybura commented:
Dear @ZeeWanderer, thank you for the comment. I'm working for support LLVM_TARGETS_TO_BUILD. Unfortunately, I'm not sure that ~28 GB will be acceptable. There are no official opinion from the CI holders.

JackBoosY committed:
Merge branch 'master' into llvm10

yurybura committed:
[llvm] add futures for main targets
[llvm] remove utils and clang-tools-extra from default features list
[llvm] exclude tests

yurybura committed:
[llvm] add features for main targets
[llvm] remove utils and clang-tools-extra from default features list
[llvm] exclude tests

yurybura committed:
Merge branch 'llvm10' of https://github.com/yurybura/vcpkg into llvm10

yurybura committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into llvm10

cbezault commented on ports/llvm/CONTROL:
```diff
@@ -1,19 +1,75 @@
 
-Feature: test
-Description: Generate build targets for the LLVM unit tests.
+Feature: target-aarch64
```
The description should probably be:

Build `architecture` backend.

yurybura replied:
Fixed. Thank you.

yurybura committed:
[llvm] update features description

yurybura committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into llvm10

yurybura commented:
Finally CI looks happy)
@strega-nil @PhoebeHui @PhoebeHui Please review/merge this PR.
Thanks!

alexreinking commented:
@yurybura -- Thank you so much for your work on this! I'm with the Halide project and (among other things) am working on improving our build on Windows. This is a huge help.

strega-nil commented:
This is extremely exciting! I'm hopeful we can merge today :3

tjysdsg commented:
Is this PR ready to be merged? I have a project that needs LLVM-10, and it would be great to vcpkg to handle this dependency! Thanks for all your hard work btw

alexreinking commented:
> Is this PR ready to be merged? I have a project that needs LLVM-10, and it would be great to vcpkg to handle this dependency! Thanks for all your hard work btw

Ditto! Happy to lend a hand any way I can to grease the skids here, too 🙂

strega-nil commented:
Ah, sorry! I got distracted and forgot to hit the button... unfortunately I'm still learning the ropes -.-

alexreinking commented:
> Ah, sorry! I got distracted and forgot to hit the button... unfortunately I'm still learning the ropes -.-

No, no, thank you so much! This is a big help to me and my team!

cenit commented:
sadly, llvm is broken on Linux now.

strega-nil commented:
@cenit looks like running out of space, so I'm gonna rerun it. This is the issue with LLVM

cenit commented:
It’s broken  also on my local machine now. Still not investigated the reason, but for sure not disk space

strega-nil commented:
That's really odd... I'm looking into it

yurybura commented:
My local build works fine. Debug build log is attached:
[install-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/4577243/install-x64-linux-dbg-out.log)

@cenit Can you attach your log?

cenit commented:
Sorry I lost logs in the update (last step is cleanup of buildtrees folder in any case). In case the same problem happens again, tonight, I will let you know. Seeing that it was broken also on master made maybe too quick to think it was this PR fault. Maybe it was just a double coincidence (two different unrelated problems here and on my machine)

cenit commented:
re-launched manually to have logs sooner:

```
/usr/bin/ar: creating t.a
CMake Error at lib/WindowsManifest/CMakeLists.txt:11 (get_filename_component):
  get_filename_component unknown component optimized


-- Version: 0.0.0
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX -- success
-- Performing Test HAVE_STEADY_CLOCK
-- Performing Test HAVE_STEADY_CLOCK
-- Performing Test HAVE_STEADY_CLOCK -- success
```

Distributor ID: Ubuntu
Description:    Ubuntu 20.04 LTS
Release:        20.04
Codename:       focal

strega-nil commented:
It's working on my machine... how do you run the build?

BillyONeal commented:
Under the hypothesis that it failed in CI due to the OOM killer, this makes the test VMs have more RAM: https://github.com/microsoft/vcpkg/pull/11174 (The PR infrastructure under which this PR passed had half as much RAM per core)

strega-nil commented:
The LLVM in the new infra is now failing in the same way as in @cenit's case. This is really, really weird.

BillyONeal commented:
Considering it turns out not to be the OOM killer and we have direct customer feedback of it being broken on their machine, should we back this out?

strega-nil commented:
Yeah, I guess so. I don't like having such an old version, but we can at least try and figure out why it's broken.

cenit commented:
For me it's ok to just wait a little bit and have llvm 10 properly installed. It just depends on timing, maybe @yurybura can help?

yurybura commented:
@cenit Your issue is related to the external dependency to the libxml2. It easy to fix, but I need full CMake log. This is CMake configuration issue.
@BillyONeal @strega-nil CI linker fail is not related to the issue mentioned above.

yurybura commented:
@cenit @BillyONeal @strega-nil I found the full log in the PR #11174. Let me try to fix it.

cenit commented:
thank you. I'd really like to help you but I am really very constrained with time... sorry

yurybura commented:
@cenit @BillyONeal @strega-nil Please review/merge my fix in the PR #11108.
Thanks a lot.
